### ISSUE TITLE
Crash when rendering a `SDL_Rect` (or `SDL_FRect`) and then some points

### ISSUE THREAD

**Issue by TerensTare** [libsdl-org/SDL#4845]
I'm trying to recreate the classic Pong game using SDL2 but I'm having a crash when using SDL's rendering functions. The "render loop" is basically:
```
1. Set the draw color to black.
2. Clear the render buffer.
3. Set the clear color to the first paddle's color.
4. Draw a rectangle representing the first paddle.
5. Repeat 3. and 4. for the second paddle.
6. Set the clear color to the color of the ball.
7. Calculate points within the ball's radius that satisfy `x ^ 2 + y ^ 2 < radius ^ 2`.
8. Render the points `(ball.x + x, ball.y + y)` for each `(x, y)` satisfying 7 (again using integer coordinates).
9. Call `SDL_RenderPresent`.
```

I have also provided the code I'm using below. When I build & run the code (both Debug & Release mode) a blank, unresponsive, frozen, white window is shown, which won't close by just pressing the X button (at least on Windows 10). Also I tried to force close the program from the task manager but it didn't. The program would then be automatically closed some time after.
the issue is still present when using floating point coordinates (if that's of help). If I comment the code representing steps 3-5 (rendering the paddles) OR the code representing steps 6-8 (rendering the ball) the program works fine. Also I wrote the provided code just for reproducing the behavior and thus I have skipped error checking 😅 .

Further notes:
OS - Windows 10
Compiler: MSVC shipped with VS 2022 Build Tools, also tested with the MSVC from latest VS 2019 Build Tools (both having the issue).
SDL2 version used: the latest one available from vcpkg.
C++ standard: 20, but the only feature used are designated initializers so the code can be easily downgraded to lower standards.

<details>
<summary>Example code that "triggers" the crash</summary>

```cpp
#include <SDL2/SDL.h>

struct Paddle final
{
    SDL_Rect bounds;
    SDL_Color color{255, 255, 255, 255};
};

struct Ball final
{
    SDL_Point pos;
    float radius;
    SDL_Color color{255, 255, 255, 255};
};

int main(int, char **)
{
    SDL_Init(SDL_INIT_EVERYTHING);

    SDL_Window *win = SDL_CreateWindow(
        "Pong",
        SDL_WINDOWPOS_UNDEFINED, SDL_WINDOWPOS_UNDEFINED,
        800, 600,
        SDL_WINDOW_SHOWN //
    );

    SDL_Renderer *ren = SDL_CreateRenderer(win, -1, SDL_RENDERER_ACCELERATED);

    Paddle paddles[2] = {
        Paddle{
            .bounds = SDL_Rect{0, 250, 10, 100},
        },
        Paddle{
            .bounds = SDL_Rect{790, 250, 10, 100},
        },
    };

    Ball ball = {
        .pos = SDL_Point{390, 290},
        .radius = 10.f,
    };

    for (bool running = true; running;)
    {
        for (SDL_Event event; SDL_PollEvent(&event);)
        {
            if (event.type == SDL_QUIT)
                running = false;
        }

        SDL_SetRenderDrawColor(ren, 0, 0, 0, 255);
        SDL_RenderClear(ren);

        SDL_SetRenderDrawColor(ren, paddles[0].color.r, paddles[0].color.g, paddles[0].color.b, paddles[0].color.a);
        SDL_RenderDrawRect(ren, &paddles[0].bounds);

        SDL_SetRenderDrawColor(ren, paddles[1].color.r, paddles[1].color.g, paddles[1].color.b, paddles[1].color.a);
        SDL_RenderDrawRect(ren, &paddles[1].bounds);

        SDL_SetRenderDrawColor(ren, ball.color.r, ball.color.g, ball.color.b, ball.color.a);

        for (int y = -10; y <= 10; ++y)
            for (int x = -10; x <= 10; ++x)
                if ((x * x + y * y) < 100)
                    SDL_RenderDrawPoint(ren, ball.pos.x + x, ball.pos.y + y);

        SDL_RenderPresent(ren);
    }

    SDL_DestroyRenderer(ren);
    SDL_DestroyWindow(win);
    SDL_Quit();

    return 0;
}
```
</details>

**Comment by slouken** [comment]
Please retest using the latest SDL snapshot at commit https://github.com/libsdl-org/SDL/commit/ea97ab61640c9f9398d4ad375c904f19d942ac6f:
http://www.libsdl.org/tmp/SDL-2.0.zip

Thanks!

**Comment by TerensTare** [comment]
Just checked and the issue still persists even with the last commit.

**Comment by slouken** [comment]
I haven't been able to reproduce this. Can you run this under the debugger and find out what's happening?

**Comment by icculus** [comment]
I can't reproduce this either. Maybe a GPU driver issue (not being able to force-quit the app suggests this is the case)...

@TerensTare, try forcing this to the software renderer, just so we can avoid Direct3D altogether, and see if the problem persists. This isn't a _solution_ to the problem, but it will give us some more insight.

Add this line before the SDL_CreateRenderer() call:

```c
SDL_SetHint(SDL_HINT_RENDER_DRIVER, "software");
```

**Comment by TerensTare** [comment]
The program runs fine if I use the software renderer.

**Comment by icculus** [comment]
Mmmm...I'm inclined to think this a driver bug.

**Comment by TerensTare** [comment]
Most probably. The drivers of the hardware I'm testing on are old enough to make me think so. I appreciate the help.

**Comment by GMHarish285** [comment]
This thread helped me a lot. I faced the same issue and spent a whole day trying to debug my game and fix any memory leak. After reading this and seeing that you have also used MSVC, I tried GNU and it worked.

I am not sure if this was a driver or compiler problem, but changing the compiler worked for me.

I also read in a really old post that Microsoft libraries might be causing memory leaks. https://discourse.libsdl.org/t/sdl-crash-on-exit/7826/3
